librad/src/identities/git.rs

```diff
@@ -0,0 +1,763 @@
+        }
+    }
+
+    /// Convenience to specialise `T` to [`Project`].
+    pub fn as_project(&self) -> Git<'_, Project> {
+        Git {
+            repo: self.repo,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Read an identity whose type is not statically known from commit `oid`.
+    ///
+    /// The only guarantee about the returned value is that it is well-formed --
+    /// it may or may not pass verification.
+    pub fn some_identity(&self, oid: git2::Oid) -> Result<SomeIdentity, error::Load> {
+        SomeIdentity::try_from(self.by_oid(oid))
+    }
+
+    /// Traverse the history with head commit `head`, yielding identities of
+    /// type `T`.
+    ///
+    /// The iterator yields elements in **reverse order**, ie. oldest-first. No
+    /// verification is performed, but `Ok` elements are guaranteed to be
+    /// well-formed.
+    pub fn iter(
+        &self,
+        head: git2::Oid,
+    ) -> Result<impl Iterator<Item = Result<T, error::Load>> + 'a, error::Load>
+    where
+        T: TryFrom<ByOid<'a>, Error = error::Load>,
+    {
+        Ok(Iter::new(self.repo, head)?
+            .map(|item: Result<generic::Verifying<T, _>, _>| item.map(|v| v.into_inner())))
+    }
+
+    //// Generic methods ////
+
+    fn get_generic(&self, oid: git2::Oid) -> Result<T, error::Load>
+    where
+        T: TryFrom<ByOid<'a>, Error = error::Load>,
+    {
+        T::try_from(self.by_oid(oid))
+    }
+
+    fn verify_generic<Doc>(
+        &self,
+        head: git2::Oid,
+    ) -> Result<VerifiedIdentity<Doc>, VerificationError>
+    where
+        Doc: Delegations + generic::Replaces<Revision = Revision>,
+        <Doc as Delegations>::Error: std::error::Error + Send + Sync + 'static,
+
+        Identity<Doc>: TryFrom<ByOid<'a>, Error = error::Load>,
+    {
+        self.fold_verify_generic(head).map(|folded| folded.head)
+    }
+
+    fn fold_verify_generic<Doc>(
+        &self,
+        head: git2::Oid,
+    ) -> Result<generic::Folded<Doc, Revision, ContentId>, VerificationError>
+    where
+        Doc: Delegations + generic::Replaces<Revision = Revision>,
+        <Doc as Delegations>::Error: std::error::Error + Send + Sync + 'static,
+
+        Identity<Doc>: TryFrom<ByOid<'a>, Error = error::Load>,
+    {
+        let mut progeny = Iter::<'_, Identity<Doc>>::new(self.repo, head)
+            .map_err(generic::error::Verify::history)?;
+
+        // TODO(kim): should we skip non-quorum commits at the beginning?
+        let root = progeny
+            .next()
+            .ok_or(generic::error::Verify::EmptyHistory)?
+            .map_err(generic::error::Verify::history)?
+            .signed()?
+            .quorum()?
+            .verified(None)?;
+
+        root.verify(progeny)
+    }
+
+    //// Helpers ////
+
+    fn by_oid(&self, oid: git2::Oid) -> ByOid<'a> {
+        (self.repo, oid)
+    }
+
+    fn is_in_ancestry_path(&self, commit: git2::Oid, tree: git2::Oid) -> Result<bool, git2::Error> {
```

Comment:
Is this different from [graph_descendant_of](https://docs.rs/git2/0.13.11/git2/struct.Repository.html#method.graph_descendant_of)?

Reply:
Yeah, it looks for the `tree` (= `Revision`), not the `commit` (= `ContentId`)